polling/frequent/activities.py

```diff
@@ -17,6 +20,16 @@ async def compose_greeting(input: ComposeGreetingInput) -> str:
     while True:
         try:
             result = test_service.get_service_result(input)
+            print(f"exiting activity ${result}")
             return result
-        except Exception:
+        except Exception as e:
+            # swallow exception since service is down
+            print(e)
+
+        try:
             activity.heartbeat("Invoking activity")
+        except CancelledError as exception:
+            # activity was either cancelled or workflow was completed or worker shut down
+            raise exception
```

Comment:
Heartbeat does not throw. Cancellation in Python occurs at the asyncio level so it will throw whatever awaits (so probably on sleep or RPC call). And it uses asyncio cancellation, so it'd be `asyncio.CancelledError` not the Temporal one. Also, there is no value in catching an exception just to re-raise (unless it's just for the doc), and if you do, just call `raise` no need to pass anything.

Reply:
I think the doc nudge to handle cancellation is merited here as it is in the samples-java repo as well. I know it has helped me to remember to deal with cancellation correctly.